nano/core_test/websocket.cpp

```diff
@@ -75,6 +75,86 @@ boost::optional<std::string> websocket_test_call (std::string host, std::string
 }
 }
 
+/** Tests clients subscribing multiple times or unsubscribing without a subscription */
+TEST (websocket, subscription_edge)
+{
+	nano::system system (24000, 1);
+	nano::node_init init1;
+	nano::node_config config;
+	nano::node_flags node_flags;
+	config.websocket_config.enabled = true;
+	config.websocket_config.port = 24078;
+
+	auto node1 (std::make_shared<nano::node> (init1, system.io_ctx, nano::unique_path (), system.alarm, config, system.work, node_flags));
+	node1->start ();
+	system.nodes.push_back (node1);
+
+	ASSERT_EQ (0, node1->websocket_server->subscriber_count (nano::websocket::topic::confirmation));
+
+	// First subscription
+	{
+		ack_ready = false;
+		std::thread subscription_thread ([]() {
+			websocket_test_call ("::1", "24078", R"json({"action": "subscribe", "topic": "confirmation", "ack": true})json", true, false);
+		});
+		subscription_thread.detach ();
+		system.deadline_set (5s);
+		while (!ack_ready)
+		{
+			ASSERT_NO_ERROR (system.poll ());
+		}
+		ASSERT_EQ (1, node1->websocket_server->subscriber_count (nano::websocket::topic::confirmation));
+	}
+
+	// Second subscription, should not increase subscriber count, only update the subscription
+	{
+		ack_ready = false;
+		std::thread subscription_thread ([]() {
+			websocket_test_call ("::1", "24078", R"json({"action": "subscribe", "topic": "confirmation", "ack": true})json", true, false);
+		});
+		subscription_thread.detach ();
+		system.deadline_set (5s);
+		while (!ack_ready)
+		{
+			ASSERT_NO_ERROR (system.poll ());
+		}
+		ASSERT_EQ (1, node1->websocket_server->subscriber_count (nano::websocket::topic::confirmation));
+	}
+
+	// First unsub
+	{
+		ack_ready = false;
+		std::thread unsub_thread ([]() {
+			websocket_test_call ("::1", "24078", R"json({"action": "unsubscribe", "topic": "confirmation", "ack": true})json", true, false);
+		});
+		unsub_thread.detach ();
+		system.deadline_set (5s);
+		while (!ack_ready)
+		{
+			ASSERT_NO_ERROR (system.poll ());
+		}
+		ASSERT_EQ (0, node1->websocket_server->subscriber_count (nano::websocket::topic::confirmation));
+	}
+
+	// Second unsub, should acknowledge but not decrease subscriber count
+	std::this_thread::sleep_for (50ms);
```

Comment:
@cryptocode this is an interesting place to test the issues we were having with websocket closing before being done, and crashing. Removing this line and running this test under Debug makes it crash every single time for me. Maybe you can get some insight?

Reply:
Yeah the sessions are outliving the node in the tests. This issue will be fixed as part of a websocket PR I'm working on.

Reply:
#2007